[0047] Add DirectX metadata schema for input/output semantic signatures (#425)
llvm/wg-hlsl: proposals/NNNN-semantic-signatures-metadata-schema.md

!4 = !{i32 0, !"SV_Position", i32 9, i32 3, !50, i32 2, i32 1, i8 4, i32 0, i8 0, i8 0, i8 0, i32 0}
!5 = !{i32 1, !"TEXCOORD",    i32 9, i32 0, !50, i32 2, i32 1, i8 4, i32 1, i8 0, i8 0, i8 0, i32 0}
!6 = !{i32 2, !"TEXCOORD",    i32 9, i32 0, !52, i32 2, i32 1, i8 4, i32 2, i8 0, i8 7, i8 0, i32 0}

; ps_main outputs
!7 = !{i32 0, !"SV_Target", i32 9, i32 16, !50, i32 0, i32 1, i8 4, i32 0, i8 0, i8 0, i8 0, i32 0}
!8 = !{i32 1, !"SV_Target", i32 9, i32 16, !52, i32 0, i32 1, i8 4, i32 1, i8 0, i8 0, i8 0, i32 0}

!50 = !{i32 0}          ; single semantic index {0}
!51 = !{i32 0, i32 1}   ; array form: one element spanning two rows, indices {0, 1}
!52 = !{i32 1}          ; single semantic index {1}
```

## Generating Metadata

The `dx.semantic.signatures` table is produced entirely from within
`CGHLSLRuntime::emitEntryFunction` in `clang/lib/CodeGen/CGHLSLRuntime.cpp`.
Briefly, this is done so by traversing through each of the entry function's
input parameters and return outputs, accumulating an in-memory representation
of the signature elements and emitting the load/store intrinsic. It uses the
order the parameters/return outputs are declared in to assign the signature id

> that will associate the semantic in the intrinsic call and the metadata table.

Reviewer: nit: 

```suggestion
of the signature elements and emitting the load/store intrinsic. Signature IDs are assigned in declaration order (parameters, then return outputs), giving each semantic a stable ID that appears in both the intrinsic call and the metadata table.
```

Reviewer: Or something similar. Just to clean up the statement a little. I had to do a double take on it when reading.

Once these have been traversed, the DXIL packing algorithm is run on the
in-memory signature elements. Finally we can emit the in-memory representation
to the metadata nodes as described above. This is described in more detail
below.

### In-Memory Representation

A `SemanticSignatureElement` struct will be defined in
`llvm/Frontend/HLSL/SemanticSignatures.h` that will contain all the operands as
above, as well as, helpers to determine the extra container field information
(eg. is allocated). It will also define functions to convert to/from the

> metadata representation and the packing algorithm is defined here.

Reviewer: nit:
```suggestion
A `SemanticSignatureElement` struct, defined in
`llvm/Frontend/HLSL/SemanticSignatures.h`, will contain all the operands as
above as well as helpers to determine the extra container field information
(eg.,  is allocated). The struct will also define functions to convert to and from the
metadata representation, and the header will define the packing algorithm.
```

### Accumulate `SemanticSignatureElement`s

`emitEntryFunction` will then have a vector of `SemanticSignatureElement`s for
both input and output signatures. The existing traversal visits each leaf

> semantic in declaration order as it generates `llvm.dx.load.input` /

Reviewer: Very minor nit, but I think its better to be super explicit and use 'and' , 'or', or whatever instead of slashes in something formal. The intent is very clear in this case as-is though.

---
title: "[NNNN] - Semantic Signature Metadata"
params:
  authors:
    - inbelic: Finn Plummer

>   status: Under Consideration

Author: ```suggestion
  status: Accepted
```

---

## Introduction

This proposal defines the LLVM IR module-level metadata schema for HLSL shader
input and output signatures when targeting DirectX, and describes how that

> metadata is generated by Clang's HLSL code generation.c

Reviewer 2: nit: typo "generation .c"

> metadata is generated by Clang's HLSL code generation.c

Reviewer: Is the '.c' a typo?

## Metadata Schema

#### Named Signature Table

```LLVM
!dx.semantic.signatures = !{ !1, !2 }
```

A named metadata node, `dx.semantic.signatures`, is used to identify the
table of per-entry-point semantic signatures. The table itself is a list of
references to function/signature triples. If no entry point in the module has
any signature, the named metadata node may be omitted entirely.

#### Function/Signature Triple

```LLVM
!1 = !{ ptr @vs_main, !3, !4 }
```

The function/signature triple associates an entry-point function with its
input and output signatures.

| # | Type          | Description |
|---|---------------|-------------|
| 0 | `ptr`         | The entry-point function these signatures belong to. |
| 1 | metadata node | Reference to input signature element list, or `null` if absent. |
| 2 | metadata node | Reference to output signature element list, or `null` if absent. |

An entry function pointer can appear at most once in `!dx.semantic.signatures`.

#### Input/Output Signature Element List

```LLVM
!3 = !{ !5, !6 }
```

A signature element list consists of a list of references to signature
element nodes.

#### Signature Element

```LLVM

> !1 = !{ i32 0, !"TEXCOORD", i32 9, i32 0, !50, i32 0, i32 1, i8 4, i32 0, i8 0, i8 0, i8 0, i32 0 }

Reviewer 3: nit: Since the previous codesnippet listed `!5` as a signature element, I think this change will make all snippets link together to make an example of a valid signature metadata representation.

```suggestion
!5 = !{ i32 0, !"TEXCOORD", i32 9, i32 0, !50, i32 0, i32 1, i8 4, i32 0, i8 0, i8 0, i8 0, i32 0 }
```

```

A signature element describes a single packed range of signature rows with
its semantic name and indices, register placement, component masks, and
stage-specific attributes. Each element must retain all relevant information to
serialize into `ISG1`, `OSG1` and `PSV0`.

| #  | Type            | Description |
|----|-----------------|-------------|
|  0 | i32             | dense 0-based index within the entry function signature list; matches the operand of `llvm.dx.load.input` / `llvm.dx.store.output` |
|  1 | metadata string | the semantic name (e.g. `!"TEXCOORD"`, `!"SV_Position"`) |

> |  2 | i32             | component type; see [`llvm::dxil::ElementType`](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/DXILABI.h). Equivalent to DXC's [`DXIL::ComponentType`](https://github.com/microsoft/DirectXShaderCompiler/blob/main/include/dxc/DXIL/DxilConstants.h); the two enums share value assignments through `PackedU8x32 = 18`. The SM 6.9 additions (`I8`, `U8`, `F8_E4M3FN`, `F8_E5M2`) will need to be mirrored into `ElementType`. |

Author: ```suggestion
|  2 | i32             | component type; see [`llvm::dxil::ElementType`](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/DXILABI.h). |
```

|  5 | i32             | interpolation mode; see [`INTERPOLATION_MODE`](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/BinaryFormat/DXContainerConstants.def) |
|  6 | i32             | number of consecutive register rows occupied |
|  7 | i8              | number of components per row (1–4) |
|  8 | i32             | starting register row; `-1` (`0xFFFFFFFF`) if unallocated |
|  9 | i8              | starting component column; `-1` (`0xFF`) if unallocated, otherwise 0–3 |
| 10 | i8              | 4-bit bitmask of components that are always read (input) or may be written (output). |
| 11 | i8              | 4-bit bitmask of components that are dynamically indexed |
| 12 | i32             | GS output stream index; 0 for non-GS stages |

The following downstream container fields are derived from the operands
above rather than carried in separate metadata operands:

- Allocated status: A signature element is allocated iff field 8
  (`StartRow`) is not `-1` (`0xFFFFFFFF`) and field 9 (`StartCol`) is not
  `-1` (`0xFF`). The two sentinels are always set together:
  `StartRow == -1` iff `StartCol == -1`.

> - Declared component mask: Computed from field 7 (`Cols`) and field 9

Reviewer: ```suggestion
- Declared component mask (DeclaredMask): Computed from field 7 (`Cols`) and field 9
```

  (`StartCol`) as `((1 << Cols) - 1) << StartCol`.

- `AlwaysReads` / `NeverWrites` Masks: Field 10 is written directly to the
  container's `AlwaysReads` mask for input signatures; for output signatures
  the container's `NeverWrites` mask is the bitwise complement of field 10

>   against the declared component mask above.

Reviewer: ```suggestion
  the container's `NeverWrites = ~UsageMask & DeclaredMask`
```

!51 = !{i32 0, i32 1}   ; array form: one element spanning two rows, indices {0, 1}
!52 = !{i32 1}          ; single semantic index {1}
```

## Generating Metadata

The `dx.semantic.signatures` table is produced entirely from within
`CGHLSLRuntime::emitEntryFunction` in `clang/lib/CodeGen/CGHLSLRuntime.cpp`.
Briefly, this is done so by traversing through each of the entry function's
input parameters and return outputs, accumulating an in-memory representation
of the signature elements and emitting the load/store intrinsic. It uses the
order the parameters/return outputs are declared in to assign the signature id
that will associate the semantic in the intrinsic call and the metadata table.
Once these have been traversed, the DXIL packing algorithm is run on the
in-memory signature elements. Finally we can emit the in-memory representation
to the metadata nodes as described above. This is described in more detail
below.

### In-Memory Representation

A `SemanticSignatureElement` struct will be defined in
`llvm/Frontend/HLSL/SemanticSignatures.h` that will contain all the operands as
above, as well as, helpers to determine the extra container field information

> (eg. is allocated). It will also define provide functions to convert to/from

Reviewer 2: nit: "It will also define provide functions ..."

> (eg. is allocated). It will also define provide functions to convert to/from

Reviewer: ```suggestion
(eg. is allocated). It will also provide functions to convert to/from
```

> the metadata representation and the packing algorithm will is defined here.

Reviewer: ```
the metadata representation and the packing algorithm will is defined here.
```

<img width="498" height="362" alt="Image" src="https://github.com/user-attachments/assets/c51d271c-a066-48cb-955d-e8c9c18cba26" />

### Accumulate `SemanticSignatureElement`s

> `emitEntryFunction` will then have a a vector of `SemanticSignatureElement`s for

Author: ```suggestion
`emitEntryFunction` will then have a vector of `SemanticSignatureElement`s for
```

vectors describe every signature element for the entry, with layout fields
left unallocated. `emitEntryFunction` then invokes the packing algorithm on the
in-memory representation.

Note that:

- Inputs and outputs are packed independently (and outputs are further
  partitioned by `GSStream` for geometry shaders).
- The packer only mutates `StartRow`, `StartCol`, `UsageMask`, and
  `DynIndexMask` on each `SemanticElement`. It never reorders, inserts, or
  removes entries, so every previously emitted call site remains valid: the
  `SigId` operand still names the same element, and the row/column operands
  are element-relative and therefore packing-agnostic.
- `UsageMask` and `DynIndexMask` are computed from the recorded uses
  observed during the traversal (which columns were read/written, and which
  accesses used a non-`poison` dynamic index operand).

### Emitting the Metadata Table

At the end of `emitEntryFunction`, with the two vectors fully populated and
packed, the metadata is materialized in one pass:

1. For each `SemanticSignatureElement`, build the operand-4 semantic-indices
   node (`!{ i32 ... }`) and the 13-operand element node described in

>    [Signature Element](#signature-element).

Author: ```suggestion
1. For each `SemanticSignatureElement`, build the semantic-indices
   node (`!{ i32 ... }`) and the signature element node described in
   [Signature Element](#signature-element).
```
